### ISSUE TITLE
Compiler bug when trying to iterate `rlock`ed array in template 

### ISSUE THREAD

**Issue by billpcs** [vlang/v#11078]
**V doctor:**
```
OS: linux, Ubuntu 18.04.5 LTS
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-3630QM CPU @ 2.40GHz
CC version: cc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0

getwd: /home/bill/Desktop/vweb_context
vmodules: /home/bill/.vmodules
vroot: /home/bill/Documents/opt/v
vexe: /home/bill/Documents/opt/v/v
vexe mtime: 2021-08-06 08:17:30
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 1e255d3.a4358a6

Git version: git version 2.17.1
Git vroot status: weekly.2021.31-33-ga4358a68
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 333c14de
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg vweb_context.v`
```v
module main

import vweb

struct App {
	vweb.Context
mut:
	nums shared []int
}

fn main() {
	app := &App{}
	vweb.run(app, 8888)
}

pub fn (mut app App) index() vweb.Result {
	data := rlock app.nums {
		app.nums
	}
	return $vweb.html()
}

```

_index.html_
```html
@for d in data
<h1> > @d </h1>
@end
```

**What did you expect to see?**

Either application executes or compiler error

**What did you see instead?**
```
==================
/tmp/v/vweb_context.5237749346280611346.tmp.c:25592: warning: assignment from incompatible pointer type
/tmp/v/vweb_context.5237749346280611346.tmp.c:25595: error: field not found: val
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

**Comment by billpcs** [comment]
Some more information:

In essence, the error is this 

```
error: ‘Array_int {aka struct array}’ has no member named ‘val’
   for (int _t2 = 0; _t2 < data->val.len; ++_t2) {
```

In file `gen/c/array.c` there is a function called `gen_array_contains` which (I think) is the one that is used for generating iterations 

This function has this logic inside that generates a `->val`

```v
if node.left_type.share() == .shared_t {
    g.write('->val')
}
```

which for some reason is not always defined. 

---

Here is a much simpler scenario:

`main.v`
``` v
fn main() {
  shared t := [1,2,3]
  arr := rlock t { t }
  $tmpl("test.txt")
}
```


`test.txt`
```
@for t in arr
  Hello @t
@end
```

**Comment by billpcs** [comment]
@spytheman can you please set the Bug label, so that this is easier to find?

**Comment by Dialga** [comment]
The problem is here:
```
-		Array_int* arr = _t1;
+		__shared_Array_int* arr = _t1;
		strings__Builder sb = strings__new_builder(90);
		strings__Builder_write_string(&sb, _SLIT(""));
		for (int _t2 = 0; _t2 < arr->val.len; ++_t2) {
			int t = ((int*)arr->val.data)[_t2];
			strings__Builder_write_string(&sb,  str_intp(2, _MOV((StrIntpData[]){{_SLIT("\n  Hello "), 0xfe07, {.d_i32 = t}}, {_SLIT0, 0, { .d_c = 0 }}})));
		}
```

**Comment by Dialga** [comment]
@yuyi98 any ideas?

**Comment by yuyi98** [comment]
```vlang
import vweb

struct App {
	vweb.Context
mut:
	nums shared []int
}

fn main() {
	app := &App{
		nums: [1, 2, 3, 4, 5, 6]
	}
	vweb.run(app, 8080)
}

pub fn (mut app App) index() vweb.Result {
	shared data := rlock app.nums {
		app.nums
	}
	return $vweb.html()
}

PS D:\Test\v\tt1> v run .
[Vweb] Running app on http://localhost:8080
```

![image](https://user-images.githubusercontent.com/6949593/140527929-42a20939-f774-43c4-8e9c-b1fa24e1bdac.png)

**Comment by Dialga** [comment]
Fixed by commit 8dc2601080ca6a8453c383c4ac93e740e2824d5b on 8th April 2022.

**Comment by Dialga** [comment]
> Here is a much simpler scenario:
> 
> `main.v`
> 
> ```v
> fn main() {
>   shared t := [1,2,3]
>   arr := rlock t { t }
>   $tmpl("test.txt")
> }
> ```
> 
> `test.txt`
> 
> ```
> @for t in arr
>   Hello @t
> @end
> ```

Fixed by 426e9d1734676bdf9e554ee1b1cc323a224095fd.
